File: data/algorithm.R
# Algorithm in R

# Algorithm Tasks: ----
# For a set of publications, determine the:
# (1) number of collaborators
# (2) collaboration types

# Load libraries ----
library(tidyverse)

# Load dataset ----
ds <- read_csv("testDataset.csv")

# Task 1 ----
ds_t1 <- ds %>% mutate(
  collabNum = case_when(
    NumberOfAuthors == 1 ~ "individual",
    NumberOfAuthors == 2 ~ "pair",
    NumberOfAuthors >= 3 ~ "group"
  ),
  .after = "PID"
)
# End Task 1


# Task 2 ----
ds_t2_prep <- ds_t1 %>%
  # Separate author column and create a new row for each author
  separate_longer_delim(Name, delim = ";") %>%

  # Separate author and affiliation
  separate_wider_delim(Name, delim = ", (", names = c("authorNameID", "affiliation"), too_few = "align_start", too_many = "merge") %>% #need to add code for multiple affiliations

  # Separate affiliation into organization and country
  separate_wider_delim(affiliation, delim = ", ", names = c("organization", "country"), too_few = "align_start", too_many = "merge") %>%

  # Remove extra ) from initial formatting
  mutate(country = str_remove_all(country, "\\)"))


ds_t2_compare <- ds_t2_prep %>%
  # Add authors column combining authorName, organization, and country data
  mutate(
    authors = str_glue_data(ds_t2_prep, "{authorNameID}, ({organization}, {country})"),
    .before = "authorNameID"
  ) %>%

  # Group rows by PID to determine collaboration type for each article
  group_by(PID) %>%

  # Add column for collaboration type & assign collaboration type:
  ## NA for individually authored publications
  ## local if all authors share organization
  ## national if all authors share country
  ## international if no shared organization or country
  mutate(
    collabType =
      if_else(NumberOfAuthors == 1, NA,
        case_when(
          n_distinct(organization) == 1 ~ "local",
          n_distinct(country) == 1 ~ "national",
          .default = "international"
        )
      ),
    .before = "collabNum"
  ) %>%

  # Combine author rows back into single row per publication
  # Discard temporary columns & duplicate rows
  mutate(
    authors = str_flatten(authors, collapse = "; ")
  ) %>%
  ungroup() %>%
  select(!authorNameID & !organization & !country) %>%
  distinct(PID, .keep_all = TRUE)
# End Task 2

File: data/algorithm.py
# Algorithm in python

# Algorithm Tasks: ----
# For a set of publications, determine the:
# (1) number of collaborators
# (2) collaboration types

# load libraries
import pandas as pd

# read dataset
ds = pd.read_csv('A2_Algorithm Design/three-by-three/data/testDataset.csv')

#print(ds.to_string())

# Task 1
ds_t1 = ds.copy()
ds_t1.insert(1, "collabNum", "")

for x in ds_t1.index:
    if ds_t1.loc[x, "NumberOfAuthors"] == 1:
        ds_t1.loc[x, "collabNum"] = "individual"
    elif ds_t1.loc[x, "NumberOfAuthors"] == 2:
        ds_t1.loc[x, "collabNum"] = "pair"
    else:
        ds_t1.loc[x, "collabNum"] = "group"

print(ds_t1)
# End Task 1

# Task 2
task2 = ds_t1.take([0,2], axis=1)

task2 = task2.join(task2["Name"].str.split(";", expand=True))

task2melt = task2.melt(id_vars=["PID","Name"], var_name="AuNum", value_name="Author", ignore_index=False)

task2melt[["Author_ID","affiliation"]] = task2melt["Author"].str.split(", \\(", expand=True)
task2melt["affiliation"] = task2melt["affiliation"].str.strip("\\)")
task2melt[["organization","country"]] = task2melt["affiliation"].str.split(",", expand=True)

task2grouped = task2melt.groupby("PID", as_index=False).nunique().take([0,4,6,7], axis=1)

task2grouped.insert(1,"collabType","")
for x in task2grouped.index:
    if task2grouped.loc[x, "Author_ID"] == 1:
        task2grouped.loc[x, "collabType"] = "no collaboration"
    elif task2grouped.loc[x, "organization"] == 1:
        task2grouped.loc[x, "collabType"] = "local"
    elif task2grouped.loc[x, "country"] == 1:
        task2grouped.loc[x, "collabType"] = "national"
    else:
        task2grouped.loc[x, "collabType"] = "international"

ds_t2 = task2grouped.take([0,1], axis = 1)

result = pd.merge(ds_t1, ds_t2, on="PID")
result.insert(1, "collabType", result.pop("collabType"))
print(result.take([0,1,2], axis=1))
# End Task 2